### PR TITLE
perf: impl tail-based tracing with minitrace

### PR DESCRIPTION
**Pull request by MrCroxx**
## What's changed and what's your intention?

> Please explain **IN DETAIL** what the changes are in this PR and why they are needed. :D

Implement tail-based tracing with [minitrace](https://github.com/tikv/minitrace-rust). Thank you, @andylokandy. 🙏

Tail-based tracing can be enabled with `mtrace` feature, the recording threshold can be configured by `.with_trace_config(trace_config: TraceConfig)` interface when building the hybrid cache. And the trace config can be dynamically modified with `.trace_config()` interface.

Enable tracing may involve some overhead. It is acceptable to locate the bottleneck for p99+ latency.

With tracing disable, there is no overhead.

Tail-based tracing with jaeger can be enabled with `foyer-bench`, too. Try this:

```bash
# if there was old result
# make monitor && make clear
make monitor
# run foyer-bench and record obtains that are longer than 5ms.
rm -rf /6536/foyer && cargo build --release --features "mtrace" && RUST_BACKTRACE=1 RUST_LOG=info ./target/release/foyer-bench --dir /6536/foyer --mem 64 --disk 102400 --file-size 64 --get-range 10000 --flushers 4 --reclaimers 4 --time 60 --writers 256 --w-rate 4 --admission-rate-limit 500 --readers 1024 --r-rate 1 --runtime --compression none --distribution zipf --distribution-zipf-s 0.8 --metrics --eviction lru --trace-obtain-ns 5000000
```

All related args with foyer-bench:

```plain
      --trace-insert-ns <TRACE_INSERT_NS>
          Record insert trace threshold. Only effective with "mtrace" feature [default: 1000000000]
      --trace-get-ns <TRACE_GET_NS>
          Record get trace threshold. Only effective with "mtrace" feature [default: 1000000000]
      --trace-obtain-ns <TRACE_OBTAIN_NS>
          Record obtain trace threshold. Only effective with "mtrace" feature [default: 1000000000]
      --trace-remove-ns <TRACE_REMOVE_NS>
          Record remove trace threshold. Only effective with "mtrace" feature [default: 1000000000]
      --trace-fetch-ns <TRACE_FETCH_NS>
          Record fetch trace threshold. Only effective with "mtrace" feature [default: 1000000000]
```

Screenshots:

image
image

## Checklist

- [x] I have written the necessary rustdoc comments
- [x] I have added the necessary unit tests and integration tests
- [x] I have passed `make all` (or `make fast` instead if the old tests are not modified) in my local environment.

### CODE REVIEW

**Comment by MrCroxx** [comment]
# Benchmarks & Examples

## main

```bash
rm -rf /6536/foyer && cargo build --release && RUST_BACKTRACE=1 RUST_LOG=info ./target/release/foyer-bench --dir /6536/foyer --mem 64 --disk 102400 --file-size 64 --get-range 10000 --flushers 4 --reclaimers 4 --time 60 --writers 256 --w-rate 4 --admission-rate-limit 500 --readers 1024 --r-rate 1 --runtime --compression none --distribution zipf --distribution-zipf-s 0.8 --metrics --eviction lru
```

## PR

```bash
rm -rf /6536/foyer && cargo build --release && RUST_BACKTRACE=1 RUST_LOG=info ./target/release/foyer-bench --dir /6536/foyer --mem 64 --disk 102400 --file-size 64 --get-range 10000 --flushers 4 --reclaimers 4 --time 60 --writers 256 --w-rate 4 --admission-rate-limit 500 --readers 1024 --r-rate 1 --runtime --compression none --distribution zipf --distribution-zipf-s 0.8 --metrics --eviction lru
```

## PR (enable `mtrace` but all records are filtered)

```bash
rm -rf /6536/foyer && cargo build --release --features "mtrace" && RUST_BACKTRACE=1 RUST_LOG=info ./target/release/foyer-bench --dir /6536/foyer --mem 64 --disk 102400 --file-size 64 --get-range 10000 --flushers 4 --reclaimers 4 --time 60 --writers 256 --w-rate 4 --admission-rate-limit 500 --readers 1024 --r-rate 1 --runtime --compression none --distribution zipf --distribution-zipf-s 0.8 --metrics --eviction lru
```

## PR (enable `mtrace` and records `obtain()`s that are longer than 5ms)

```bash
rm -rf /6536/foyer && cargo build --release --features "mtrace" && RUST_BACKTRACE=1 RUST_LOG=info ./target/release/foyer-bench --dir /6536/foyer --mem 64 --disk 102400 --file-size 64 --get-range 10000 --flushers 4 --reclaimers 4 --time 60 --writers 256 --w-rate 4 --admission-rate-limit 500 --readers 1024 --r-rate 1 --runtime --compression none --distribution zipf --distribution-zipf-s 0.8 --metrics --eviction lru --trace-obtain-ns 5000000
```

## Results (in 1 pic)

image

File: foyer/src/hybrid/cache.rs [hunk 1]
```diff
@@ -12,21 +12,46 @@
...
 use super::writer::HybridCacheStorageWriter;
 
+macro_rules! try_cancel {
+    ($self:ident, $span:ident, $threshold:ident) => {
+        if let Some(elapsed) = $span.elapsed() {
+            // println!(
+            //     "threshold: {}, op: {}",
+            //     $self.trace_config.$threshold.load(Ordering::Relaxed),
+            //     elapsed.as_nanos() as usize,
+            // );
```
Comment (hzxa21) [inline foyer/src/hybrid/cache.rs:46]: remove?

File: foyer/src/hybrid/cache.rs [hunk 2]
```diff
@@ -12,21 +12,46 @@
...
+            if (elapsed.as_nanos() as usize) < $self.trace_config.$threshold.load(Ordering::Relaxed) {
+                $span.cancel();
+                // println!("cancel!");
```
Comment (hzxa21) [inline foyer/src/hybrid/cache.rs:49]: remove?

**Review by hzxa21 — APPROVED** [review]
Generally LGTM

**Pull request merged**